### PR TITLE
Issue #26 real latency data

### PR DESCRIPTION
**Pull request by bkolad**

### CODE REVIEW

File: src/main/java/net/consensys/wittgenstein/core/NetworkLatency.java [hunk 1]
```diff
@@ -81,6 +83,32 @@ public int getLatency(Node from, Node to, int delta) {
...
+
+    public NetworkLatencyByCity(CSVLatencyReader csvLatencyReader) {
+      this.latencyMatrix = csvLatencyReader.getLatencyMatrix();
+    }
+
+    public int getLatency(Node from, Node to, int delta) {
+
+      String cityFrom = from.cityName.get();
+      String cityTo = to.cityName.get();
+
```
Comment (bkolad) [inline src/main/java/net/consensys/wittgenstein/core/NetworkLatency.java]: Shall we divide it by 2? The measurements are for round trip time but we care about one way, the same question for other latency models

File: src/main/java/net/consensys/wittgenstein/core/Node.java [hunk 2]
```diff
@@ -37,6 +40,7 @@
   protected long msgSent = 0;
   protected long bytesSent = 0;
   protected long bytesReceived = 0;
+  public final Optional<String> cityName;
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/core/Node.java]: I think it's safer to have a global city by default, this save the optional part.
The current code does not use optional, some fields can be null, but optional is never used. For consistency it's better not use optional (or then do a global change)...

File: src/main/java/net/consensys/wittgenstein/protocol/GSFSignature.java [hunk 3]
```diff
@@ -452,8 +466,8 @@ public String toString() {
 
   @Override
   public Protocol copy() {
-    return new GSFSignature(nodeCount, threshold, pairingTime, timeoutPerLevelMs, periodDurationMs,
-        acceleratedCallsCount, nodesDown, this.network.networkLatency);
+    return new GSFSignature(nb, network, nodeCount, threshold, pairingTime, timeoutPerLevelMs,
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/protocol/GSFSignature.java]: For a copy you want to use a new network (the network object is 'used' when running a scenario.

File: src/main/java/net/consensys/wittgenstein/tools/CSVLatencyReader.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  public static void main(String[] args) throws IOException {
+    CSVLatencyReader reader = new CSVLatencyReader();
+    Map<String, Map<String, Float>> latencyMatrix = reader.latencyMatrix;
+
+    Set<String> cities = latencyMatrix.keySet();
+
+    for (String cityFrom : cities) {
+
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/tools/CSVLatencyReader.java]: why all these empty lines?

File: src/test/java/net/consensys/wittgenstein/protocol/GSFSignatureTest.java [hunk 5]
```diff
@@ -1,11 +1,12 @@
 package net.consensys.wittgenstein.protocol;
 
+import net.consensys.wittgenstein.core.Network;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
 public class GSFSignatureTest {
-  private GSFSignature p = new GSFSignature(32, 1, 3, 20, 10, 10, 0);
+  private GSFSignature p = new GSFSignature(new Network(), 32, 1, 3, 20, 10, 10, 0);
```
Comment (nkeywal) [inline src/test/java/net/consensys/wittgenstein/protocol/GSFSignatureTest.java]: new Network<>();

File: src/main/java/net/consensys/wittgenstein/protocol/GSFSignature.java [hunk 6]
```diff
@@ -60,16 +62,28 @@ public GSFSignature(int nodeCount, int threshold, int pairingTime, int timeoutPe
...
-    this.nb = new Node.NodeBuilderWithRandomPosition(network.rd);
+    this.network = network;
+    this.nb = nb;
     this.network.setNetworkLatency(nl);
   }
 
-  public GSFSignature(int nodeCount, double ratioThreshold, int pairingTime, int timeoutPerLevelMs,
-      int periodDurationMs, int acceleratedCallsCount, double ratioNodesDown) {
-    this(nodeCount, (int) (ratioThreshold * nodeCount), pairingTime, timeoutPerLevelMs,
+  public GSFSignature(Network network, int nodeCount, double ratioThreshold, int pairingTime,
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/protocol/GSFSignature.java]: Why do we need this?

File: src/main/java/net/consensys/wittgenstein/protocol/GSFSignature.java [hunk 7]
```diff
@@ -527,7 +569,7 @@ public static void timePerDead() {
     long startAt = System.currentTimeMillis();
     StatsHelper.SimpleStats s;
     for (double dead = 0; dead < 0.5; dead += 0.01) {
-      ps1 = new GSFSignature(nodeCt, 1 - dead, 3, toL, pd, 10, dead);
+      ps1 = new GSFSignature(new Network(), nodeCt, 1 - dead, 3, toL, pd, 10, dead);
       ps1.network.setNetworkLatency(nl);
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/protocol/GSFSignature.java]: We don't have the new latency object in this commit?

File: src/main/java/net/consensys/wittgenstein/core/Node.java (hunk 2 above)
Comment (bkolad) [reply src/main/java/net/consensys/wittgenstein/core/Node.java]: I will change it to default value

**Pull request merged**